atomic_reactor/plugins/pre_pin_operator_digest.py

```diff
@@ -262,6 +291,90 @@ def _get_pullspecs(self, operator_csv, skip_all):
         return pullspecs
 
     def _get_replacement_pullspecs(self, pullspecs):
+        """Replace components of pullspecs
+
+        :param pullspecs: a list of pullspecs.
+        :type pullspecs: list[ImageName]
+        :return: a list of replacement result. Each of the replacement result
+            is a mapping containing key/value pairs:
+
+            * ``original``: ImageName, the original pullspec.
+            * ``new``: ImageName, the replaced/non-replaced pullspec.
+            * ``pinned``: bool, indicate whether the tag is replaced with a
+                          specific digest.
+            * ``replaced``: bool, indicate whether the new pullspec has change
+                            of repository or registry.
+
+        :rtype: list[dict[str, ImageName or bool]]
+        :raises RuntimeError: if pullspecs cannot be properly replaced
+        """
+        if self.operator_csv_modifications:
+            replacements = self._get_replacement_pullspecs_from_csv_modifications(pullspecs)
+        else:
+            replacements = self._get_replacement_pullspecs_OSBS_resolution(pullspecs)
+
+        replacement_lines = "\n".join(
+            "{original} -> {new}".format(**r) if r['replaced']
+            else "{original} - no change".format(**r)
+            for r in replacements
+        )
+        self.log.info("To be replaced:\n%s", replacement_lines)
+
+        return replacements
+
+    def _get_replacement_pullspecs_from_csv_modifications(self, pullspecs):
+        """Replace components of pullspecs based on externally provided CSV modifications
+
+        :param pullspecs: a list of pullspecs.
+        :type pullspecs: list[ImageName]
+        :return: a list of replacement result. Each of the replacement result
+            is a mapping containing key/value pairs:
+
+            * ``original``: ImageName, the original pullspec.
+            * ``new``: ImageName, the replaced/non-replaced pullspec.
+            * ``pinned``: bool, indicate whether the tag is replaced with a
+                          specific digest.
+            * ``replaced``: bool, indicate whether the new pullspec has change
+                            of repository or registry.
+
+        :rtype: list[dict[str, ImageName or bool]]
+        :raises RuntimeError: if provided CSV modification doesn't contain all
+                              required pullspecs or contain different ones
+        """
+        mod_pullspec_repl = self.operator_csv_modifications.get('pullspec_replacements', [])
+
+        # check if modification data contains all required pullspecs
+        pullspecs_set = set(pullspecs)
+        mod_pullspecs_set = set((ImageName.parse(p['original']) for p in mod_pullspec_repl))
```

Comment:
you should also verify that in provided replacements aren't duplicate 'original' entries with different other values

YOU might ignore duplicates IF all keys for duplicate entry are the same,
but I'd still say there shouldn't be any duplicates in provided replacements

Reply:
I'd address this in PR where JSONschema validation will be added